### PR TITLE
Add initial MCore integration for MFSDP v2

### PR DESCRIPTION
**Pull request by wujingyue**
## Summary

Related to #5614.

This PR adds the initial MCore integration path for experimental MFSDP v2:

- expose `FullyShardedDataParallel` as the MFSDP adapter alias selected by `MFSDP_VERSION=2`
- wrap configured FSDP unit modules before the root module for MFSDP v2
- route MFSDP v2 optimizer setup through MCore `DistributedOptimizer`
- use MFSDP v2 DTensor parameter/gradient shards directly instead of V1 param-and-grad buffers
- disable the V1 startup parameter all-gather knob when `MFSDP_VERSION=2`, since V2 gathers parameters from module hooks
- filter empty local DTensor shards from optimizer param groups while TransformerEngine FusedAdam handling is fixed upstream
- fail fast for unsupported V2 configuration paths, including loss scaling, FP16, V1-only FSDP tuning options, HSDP, FP8/FP4, and unsupported optimizer features
- add distributed unit coverage against an unsharded MCore mixed-precision optimizer reference
- route the H100 unit-test recipe’s MFSDP v2 bucket through `MFSDP_VERSION=2`

## Notes

MFSDP v2 optimizer checkpointing is still intentionally unsupported in this initial path. See #5534.

The empty-shard optimizer filtering is tied to NVIDIA/TransformerEngine#3207.

## Validation

Ran a 5-iteration `pretrain_hybrid.py` DP=2 smoke with the same mock-data/model settings for no FSDP, MFSDP v1, and MFSDP v2. The MFSDP commands pass `--ckpt-format fsdp_dtensor` only because current Megatron-FSDP argument validation requires it; this smoke does not save or load checkpoints.

```bash
COMMON_ARGS=(
  pretrain_hybrid.py
  --mock-data
  --tokenizer-type NullTokenizer
  --vocab-size 1024
  --null-tokenizer-eod-id 0
  --null-tokenizer-pad-id 1
  --bf16
  --no-gradient-accumulation-fusion
  --tensor-model-parallel-size 1
  --pipeline-model-parallel-size 1
  --context-parallel-size 1
  --expert-model-parallel-size 1
  --num-layers 2
  --hidden-size 64
  --ffn-hidden-size 128
  --num-attention-heads 4
  --group-query-attention
  --num-query-groups 4
  --mamba-num-heads 8
  --mamba-head-dim 8
  --mamba-num-groups 8
  --hybrid-layer-pattern 'M*'
  --spec megatron.core.models.hybrid.hybrid_layer_specs hybrid_stack_spec
  --seq-length 16
  --max-position-embeddings 16
  --micro-batch-size 1
  --global-batch-size 2
  --train-iters 5
  --eval-iters 0
  --eval-interval 100000
  --save-interval 100000
  --log-interval 1
  --lr 1e-4
  --min-lr 1e-5
  --lr-decay-style constant
  --weight-decay 0.0
  --clip-grad 0.0
  --attention-backend unfused
  --transformer-impl transformer_engine
  --distributed-backend nccl
)

# no FSDP
CUDA_VISIBLE_DEVICES=0,1 UV_NO_SYNC=1 \
  uv run python -m torch.distributed.run --nproc-per-node 2 \
  "${COMMON_ARGS[@]}" \
  --use-distributed-optimizer

# MFSDP v1
env -u MFSDP_VERSION CUDA_VISIBLE_DEVICES=0,1 UV_NO_SYNC=1 \
  uv run python -m torch.distributed.run --nproc-per-node 2 \
  "${COMMON_ARGS[@]}" \
  --use-megatron-fsdp \
  --use-distributed-optimizer \
  --ckpt-format fsdp_dtensor

# MFSDP v2
MFSDP_VERSION=2 CUDA_VISIBLE_DEVICES=0,1 UV_NO_SYNC=1 \
  uv run python -m torch.distributed.run --nproc-per-node 2 \
  "${COMMON_ARGS[@]}" \
  --use-megatron-fsdp \
  --use-distributed-optimizer \
  --ckpt-format fsdp_dtensor
```

Logged rank-0 `lm loss`:

| iteration | no FSDP | MFSDP v1 | MFSDP v2 |
| ---: | ---: | ---: | ---: |
| 1 | 6.945961 | 6.945961 | 6.945961 |
| 2 | 6.946891 | 6.946891 | 6.946891 |
| 3 | 6.958145 | 6.958147 | 6.958147 |
| 4 | 6.924651 | 6.924606 | 6.924606 |
| 5 | 6.908551 | 6.908510 | 6.908510 |

No skipped or NaN iterations were reported in any run.

### CODE REVIEW

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 1]
```diff
@@ -483,6 +491,182 @@ def sync_rng_states_across_tp_group(self):
...
+        if device is None:
+            device = next(module.parameters()).device
+        dp_group = pg_collection.dp_cp
+        mesh = DeviceMesh.from_group(dp_group, device_type=device.type, mesh_dim_names=("dp",))
+        placements = Placements(
+            dp_axes=[0], parameter=[Flat()], gradient=[Flat()], optimizer=[Flat()]
+        )
+        for submodule in reversed(list(module.modules())):
+            if submodule is module:
+                continue
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:546]: why is this needed?

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 2]
```diff
@@ -483,6 +491,182 @@ def sync_rng_states_across_tp_group(self):
...
+            raise ValueError("MFSDP v2 does not support disabling bucketing.")
+        if not hasattr(pg_collection, 'dp_cp'):
+            raise ValueError("MFSDP v2 requires an explicit dp_cp process group.")
+
+        parallel_sizes = {
+            "tensor model parallelism": config.tensor_model_parallel_size,
+            "pipeline model parallelism": config.pipeline_model_parallel_size,
+            "context parallelism": config.context_parallel_size,
+            "expert model parallelism": config.expert_model_parallel_size,
+        }
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py]: how about 

```
parallel_size_attributes = ["tensor_model_parallel_size", ...]
```

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -679,6 +680,12 @@ def __init__(
 
         self.is_stub_optimizer = False
         if self.ddp_config.use_megatron_fsdp:
+            if self._uses_mfsdp_v2():
+                # V1 sets this marker inside MegatronFSDP's param-and-grad buffer setup. V2
+                # bypasses that buffer path, so mark optimizer params here for generic MCore
+                # gradient statistics to consume the local shard of MFSDP DTensor gradients.
+                for parameter in self.get_parameters():
+                    parameter.__fsdp_param__ = True
```
Comment (shjwudp) [inline megatron/core/optimizer/distrib_optimizer.py]: nit: `__fsdp_param__` is used by TE’s wgrad fusion. Moving it into the `fully_shard` interface would enable us to optimize TE wgrad fusion in non-MCore scenarios.

**Review by shjwudp — APPROVED** [review]

**Comment by zhongbozhu** [comment]
One general comment: do we plan to fully replace V1 so that users don't have to worry about V1 V2?

**Comment by wujingyue** [comment]
> One general comment: do we plan to fully replace V1 so that users don't have to worry about V1 V2?

Yes, @zhongbozhu. Refer to http://nv/mfsdp-design => "implementation plan"
